### PR TITLE
Replace PARALLEL_FOR1 and PARALLEL_FOR2 with PARALLEL_FOR_IF.

### PR DESCRIPTION
**Pull request by quantumsteve**
Description of work.

This replaces `PARALLEL_FOR1` and `PARALLEL_FOR2` with `PARALLEL_FOR_IF` and the variadic function ThreadSafe(...) to determine whether one or more workspaces thread-safe.

**To test:**

This follows up on #17145.

_Does not need to be in the release notes._

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataHandling/src/LoadEventNexus.cpp [hunk 1]
```diff
@@ -1614,7 +1614,7 @@ void LoadEventNexus::loadEvents(API::Progress *const prog,
       if (mT0 != 0.0) {
         int64_t numHistograms =
             static_cast<int64_t>(m_ws->getNumberHistograms());
-        PARALLEL_FOR1(m_ws)
+        PARALLEL_FOR_IF(m_ws->threadSafe())
```
Comment (quantumsteve) [inline Framework/DataHandling/src/LoadEventNexus.cpp]: `m_ws` is an `EventWorkspaceCollection` which does NOT inherit from `DataItem` and `m_ws->threadsafe() will always return true. Is there any reason to keep this check?

**Comment by quantumsteve** [comment]
@martyngigg Would you mind looking at this next week (Before the next review meeting)?

File: Framework/DataHandling/src/LoadEventNexus.cpp (hunk 1 above)
Comment (martyngigg) [inline Framework/DataHandling/src/LoadEventNexus.cpp]: Is it worth keeping for consistency with the the general style in many other algorithms?

**Comment by martyngigg** [comment]
Apart from the comment raised around the `EventWorkspaceCollection` in LoadEventNexus this looks good.

File: Framework/DataHandling/src/LoadEventNexus.cpp (hunk 1 above)
Comment (quantumsteve) [inline Framework/DataHandling/src/LoadEventNexus.cpp]: Is it worth maintaining the static_assert that the workspace inherits from DataItem?

Comment (martyngigg) [inline Framework/DataHandling/src/LoadEventNexus.cpp]: Maybe it's better to remove that and just require a pointer type with a `threadSafe()` method?

**Comment by martyngigg** [comment]
@quantumsteve The branch has developer a conflict

**Comment by quantumsteve** [comment]
@martyngigg I removed the `static_assert` that was incompatible with `EventWorkspaceCollection`.

**Comment by martyngigg** [comment]
This looks good now. It provides a much cleaner macro for the OMP loops.

:shipit:

**Pull request merged**